csharp/src/Google.Protobuf/WellKnownTypes/TimestampPartial.cs

```diff
@@ -222,6 +222,93 @@ internal static string ToJson(long seconds, int nanoseconds, bool diagnosticOnly
             }
         }
 
+        /// <summary>
+        /// Given another timestamp, returns 0 if the timestamps are equivalent, -1 if this timestamp precedes the other, and 1 otherwise
+        /// </summary>
+        /// <param name="other">Timestamp to compare</param>
+        /// <returns>an integer indicating whether this timestamp precedes or follows the other</returns>
+        public int CompareTo(Timestamp other)
+        {
+            if (!IsNormalized(Seconds, Nanos))
+                return Normalize(Seconds, Nanos).CompareTo(other);
+            if (other == null)
```

Comment:
Better put null check in the first place (before normalize this timestamp). 
Why compare to null should return 1?  Is any other well known compare do like this that consider null as the smallest?

Reply:
The BCL treats null this way.  Consider: ```Debug.Assert(1 == "bcl".CompareTo(null));```

See my other comment on normalization and how I think we should reconsider not doing it, but if we keep it, here are my thoughts here:

Doing a null check first will do the null check twice.  Since the only Normalize() function we have creates an entirely new Timestamp instance, we just call that one's CompareTo() which will be doing its own null check.

I would understand if you wanted to do it anyway, considering un-normalized to be the exceptional case, and normalization to be the more expensive operation.  So let me know.  But I recommend we just not normalize in the first place.